src/custom_component_renderer/book_directory_renderer.rs

```diff
@@ -0,0 +1,280 @@
+                relative_path = without_lang.to_owned();
+            }
+            if Some(identifier) != i18_config.default_language.as_ref() {
+                relative_path = Path::new(identifier).join(relative_path).to_owned();
+            }
+            language_to_rendered_path.insert(
+                identifier.clone(),
+                Path::new("/").join(relative_path).to_owned(),
+            );
+        }
+        Ok(RenderingContext {
+            path,
+            language,
+            i18_config,
+            language_to_rendered_path,
+        })
+    }
+}
+
+pub(crate) struct BookDirectoryRenderer {
+    config: I18nConfiguration,
+    book: mdbook::MDBook,
+    book_dir: PathBuf,
+    components: Vec<CustomComponent>,
+    languages_paths: BTreeMap<String, PathBuf>,
+}
+
+impl BookDirectoryRenderer {
+    pub(crate) fn new(
+        config: I18nConfiguration,
+        book: mdbook::MDBook,
+        book_dir: PathBuf,
+    ) -> BookDirectoryRenderer {
+        let default_language = config.default_language.clone();
+        let languages_paths = config
+            .languages
+            .keys()
+            .filter(|language| {
+                default_language.is_none() || *language != default_language.as_ref().unwrap()
+            })
+            .map(|language| (language.clone(), book_dir.join("html").join(language)))
+            .collect::<BTreeMap<String, PathBuf>>();
+        BookDirectoryRenderer {
+            config,
+            book,
+            languages_paths,
+            book_dir,
+            components: Vec::new(),
+        }
+    }
+
+    pub fn translate(&mut self) -> Result<()> {
+        let default_language = &self.config.default_language;
+        let original_language = self.book.config.book.language.clone();
+        let book_dir = self.book_dir.as_path();
+
+        for identifier in self.config.languages.keys() {
+            if let Some(default_language) = default_language {
+                if default_language == identifier {
+                    continue;
+                }
+            }
+
+            let translation_path = book_dir.join(identifier);
+
+            self.book.config.book.language = Some(identifier.clone());
+            self.book.config.book.multilingual = true;
+            self.book.config.build.build_dir = translation_path;
+            self.book.build()?;
+            std::fs::rename(
+                book_dir.join(identifier).join("html"),
+                book_dir.join("html").join(identifier),
+            )?;
+        }
+        self.book.config.book.language = original_language;
+        self.book.config.build.build_dir = book_dir.to_owned();
+        Ok(())
+    }
+
+    pub(crate) fn render_book(&mut self) -> Result<()> {
+        let html_dir = self.book_dir.join("html");
+        if !html_dir.is_dir() {
+            return Err(RendererError::InvalidPath(format!(
+                "{:?} is not a directory",
```

Comment:
Are you assuming here that people first run `mdbook build` with the normal `html` renderer?

Should you not call the renderer yourself so that you don't need to error out here?

Reply:
This is still an important question: we cannot rely on "chance" here :slightly_smiling_face:

Reply:
Can't we simply define that as invalid and tell people they need to enable the HTML renderer? mdbook-pdf does that

Reply:
My concern is the ordering: if the HTML renderer runs after this renderer, what kind of output do you get?

Let us imagine you run `mdbook build` $n$ times. I am afraid that the new renderer will use the HTML output from run $n-1$ when it produces output for run $n$. That will put us into the same annoying situation as when using LaTeX where you need to compile 2-3 times to be "really sure" that the output is stable.

Reply:
No, it's guaranteed to run in that order with the current implementation because mdbook calls the renderers in alphabetical order. What's not ideal is that it's not documented, so it's Hyrum law for now. I will open a PR to document that behaviour when we submit this PR.

Reply:
I'd rather even add a `after` parameter in their `book.toml` params

Reply:
> I will open a PR to document that behaviour when we submit this PR.

It's great that you checked this — please open the PR already now.

We would also need to document this very carefully in our own code so that we know that the naming is important :grimacing: